lib/hash.h

```diff
@@ -2,42 +2,301 @@
+/// `hash_combine` / `hash_combine_generic` functions.
+template <class Key, class Enable = void>
+struct Hasher;
+
+struct Hash {
+    template <class T>
+    constexpr size_t operator()(const T &v) const noexcept(noexcept(Hasher<T>()(v))) {
+        return Hasher<T>()(v);
+    }
+
+    template <class T, class... Types>
+    constexpr size_t operator()(const T &t, const Types &...ts) const {
+        return hash_combine((*this)(t), (*this)(ts...));
+    }
+
+    constexpr size_t operator()() const noexcept { return 0; }
+};
+
+template <>
+struct Hasher<bool> {
+    constexpr size_t operator()(bool val) const noexcept {
+        return val ? std::numeric_limits<size_t>::max() : 0;
+    }
+};
+
+template <>
+struct Hasher<unsigned long long> : Detail::IntegerHasher<unsigned long long> {};
+
+template <>
+struct Hasher<signed long long> : Detail::IntegerHasher<signed long long> {};
+
+template <>
+struct Hasher<unsigned long> : Detail::IntegerHasher<unsigned long> {};
+
+template <>
+struct Hasher<signed long> : Detail::IntegerHasher<signed long> {};
+
+template <>
+struct Hasher<unsigned int> : Detail::IntegerHasher<unsigned int> {};
+
+template <>
+struct Hasher<signed int> : Detail::IntegerHasher<signed int> {};
+
+template <>
+struct Hasher<unsigned short> : Detail::IntegerHasher<unsigned short> {};
+
+template <>
+struct Hasher<signed short> : Detail::IntegerHasher<signed short> {};
+
+template <>
+struct Hasher<unsigned char> : Detail::IntegerHasher<unsigned char> {};
+
+template <>
+struct Hasher<signed char> : Detail::IntegerHasher<signed char> {};
+
+template <>
+struct Hasher<char> : Detail::IntegerHasher<char> {};
+
+template <>
+struct Hasher<float> : Detail::FloatHasher<float> {};
+
+template <>
+struct Hasher<double> : Detail::FloatHasher<double> {};
+
+template <>
+struct Hasher<std::string> {
+    size_t operator()(const std::string &val) const {
+        return static_cast<size_t>(hash(val.data(), val.size()));
+    }
+};
+
+template <>
+struct Hasher<std::string_view> {
+    size_t operator()(const std::string_view &val) const {
+        return static_cast<size_t>(hash(val.data(), val.size()));
+    }
+};
+
+template <typename T1, typename T2>
+struct Hasher<std::pair<T1, T2>> {
+    size_t operator()(const std::pair<T1, T2> &val) const { return Hash()(val.first, val.second); }
+};
+template <typename... Types>
+struct Hasher<std::tuple<Types...>> {
+    size_t operator()(const std::tuple<Types...> &val) const { return apply(Hash(), val); }
+};
+
+/// In general, pointers are bad hashes: their low 2-3 bits are zero, likewise
+/// for the upper bits depending on the ABI. Also, the middle bits might not
+/// have enough entropy as addresses come from some common pool. To solve this
+/// problem we just use a single iteration of hash_avalanche to improve mixing
+/// (see Detail::IntegerHasher).
+template <typename T>
+struct Hasher<T *> {
+    // FIXME: better use std::bit_cast from C++20
```

Comment:
I am thinking we should create an umbrella issue that tracks this as part of a todo list. We are not far from C++20.